RxJavaTemplate/app/src/androidTest/java/co/nimblehq/rxjava/ui/screens/home/HomeFragmentTest.kt

```diff
@@ -0,0 +1,159 @@
+import androidx.recyclerview.widget.RecyclerView
+import androidx.test.core.app.ActivityScenario
+import androidx.test.core.app.ActivityScenario.launch
+import androidx.test.espresso.Espresso.onView
+import androidx.test.espresso.Espresso.pressBack
+import androidx.test.espresso.IdlingRegistry
+import androidx.test.espresso.action.ViewActions.click
+import androidx.test.espresso.action.ViewActions.scrollTo
+import androidx.test.espresso.assertion.ViewAssertions.matches
+import androidx.test.espresso.contrib.RecyclerViewActions.actionOnItemAtPosition
+import androidx.test.espresso.matcher.ViewMatchers.*
+import androidx.test.espresso.web.assertion.WebViewAssertions.webMatches
+import androidx.test.espresso.web.model.Atoms.getCurrentUrl
+import androidx.test.espresso.web.sugar.Web.onWebView
+import co.nimblehq.rxjava.R
+import co.nimblehq.rxjava.di.modules.RepositoryModule
+import co.nimblehq.rxjava.domain.repository.ApiRepository
+import co.nimblehq.rxjava.ui.common.IdlingResource
+import co.nimblehq.rxjava.ui.screens.MainActivity
+import dagger.hilt.android.testing.HiltAndroidRule
+import dagger.hilt.android.testing.HiltAndroidTest
+import dagger.hilt.android.testing.UninstallModules
+import org.hamcrest.CoreMatchers.allOf
+import org.hamcrest.CoreMatchers.containsString
+import org.hamcrest.Matchers.not
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import javax.inject.Inject
+
+@UninstallModules(RepositoryModule::class)
+@HiltAndroidTest
+class HomeFragmentTest {
+
+    @get:Rule
+    var hiltRule = HiltAndroidRule(this)
+
+    @Inject
+    lateinit var testApiRepository: ApiRepository
+
+    @Before
+    fun setup() {
+        hiltRule.inject()
+        launchActivity()
+    }
+
+    @Test
+    fun when_initialize_should_shows_ui_correctly() {
+        onView(withId(R.id.rvHomeData)).check(matches(isDisplayed()))
+        onView(withId(R.id.pbLoading)).check(matches(isDisplayed()))
+        onView(withId(R.id.btHomeRefresh)).check(matches(not(isEnabled())))
+
+        registerIdlingResource()
+
+        onView(withId(R.id.pbLoading)).check(matches(not(isDisplayed())))
+        onView(withId(R.id.btHomeRefresh)).check(matches(isEnabled()))
+        onView(allOf(withId(R.id.tvDataTitle), withText("title1"))).check(matches(isDisplayed()))
+        onView(allOf(withId(R.id.tvDataAuthor), withText("author1"))).check(matches(isDisplayed()))
+        onView(allOf(withId(R.id.tvDataTitle), withText("title2"))).check(matches(isDisplayed()))
+        onView(allOf(withId(R.id.tvDataAuthor), withText("author2"))).check(matches(isDisplayed()))
+        onView(allOf(withId(R.id.tvDataTitle), withText("title3"))).check(matches(isDisplayed()))
+        onView(allOf(withId(R.id.tvDataAuthor), withText("author3"))).check(matches(isDisplayed()))
+
+        unregisterIdlingResource()
+    }
+
+    @Test
+    fun when_clicking_on_refresh_button_should_refresh_data() {
+        registerIdlingResource()
+
+        unregisterIdlingResource()
```

Comment:
redundant call?

Reply:
No, The point of this is to test how the app handles the loading state. When we register IdlingResource, the test will stop running until there's no IdlingResource left, which this case is when we finished getting data. After that, we unregister IdlingResource which means we don't care about delay time when getting data anymore. If we don't unregister it at that point when we start clicking on the refresh button the test will hang until the data is finished loading. At that time the progress bar should be disappeared already which means we unable to test the progress bar while it's showing. Anyway, if you can think of a less obscure way please tell me. 😄